sdk/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+                        lo_lo[7], lo_hi[0], lo_hi[1], lo_hi[2], lo_hi[3], lo_hi[4], lo_hi[5],
+                        lo_hi[6], lo_hi[7], hi_lo[0], hi_lo[1], hi_lo[2], hi_lo[3], hi_lo[4],
+                        hi_lo[5], hi_lo[6], hi_lo[7], hi_hi[0], hi_hi[1], hi_hi[2], hi_hi[3],
+                        hi_hi[4], hi_hi[5], hi_hi[6], hi_hi[7],
+                    ]
+                }
+            }
+
+            impl Display for FieldElement {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    write!(f, "0x{}", hex::encode(self.to_bytes()))
+                }
+            }
+        }
+    }
+
+    pub mod stream {
+        pub mod v1alpha2 {
+            tonic::include_proto!("apibara.node.v1alpha2");
+
+            pub(crate) const FILE_DESCRIPTOR_SET: &[u8] =
+                tonic::include_file_descriptor_set!("starknet_descriptor_v1alpha2");
+
+            pub fn stream_file_descriptor_set() -> &'static [u8] {
+                FILE_DESCRIPTOR_SET
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Configuration {
+    batch_size: u64,
+    starting_cursor: Option<Cursor>,
+    finality: Option<i32>,
+    filter: Option<Filter>,
+}
+impl Configuration {
+    pub fn new(
+        batch_size: u64,
+        starting_cursor: Option<Cursor>,
+        finality: Option<i32>,
+        filter: Option<Filter>,
+    ) -> Self {
+        Self {
+            batch_size,
+            starting_cursor,
+            finality,
+            filter,
+        }
+    }
+}
+impl Default for Configuration {
+    fn default() -> Self {
+        Self {
+            batch_size: 1,
+            starting_cursor: None,
+            finality: None,
+            filter: Some(Filter::default()),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ApibaraIndexerError {
+    #[error("Failed to build indexer")]
+    FailedToBuildIndexer,
+    #[error(transparent)]
+    TonicError(#[from] tonic::transport::Error),
+    #[error(transparent)]
+    FailedToConfigureStream(#[from] SendError<Configuration>),
+    #[error(transparent)]
+    StreamError(#[from] tonic::Status),
+}
+
+pub struct ApibaraIndexer<StreamClient> {
+    client: Option<StreamClient>,
+    configuration: Option<Configuration>,
+    configuration_channel: Option<(Sender<Configuration>, Receiver<Configuration>)>,
+    inner_channel: Option<(Sender<StreamDataRequest>, Receiver<StreamDataRequest>)>,
+}
+
+impl<StreamClient> Default for ApibaraIndexer<StreamClient> {
+    fn default() -> Self {
+        Self {
+            client: None,
+            configuration: None,
+            configuration_channel: None,
+            inner_channel: None,
+        }
+    }
+}
+
+pub struct StreamHandler {
+    stream_id: u64,
+    configuration_rx: Receiver<Configuration>,
+    inner: StreamingClientWrapper,
+    inner_tx: Sender<StreamDataRequest>,
+}
+
+#[derive(Debug)]
+#[pin_project]
+struct StreamingClientWrapper {
```

Comment:
Do we really need the extra struct? Or `inner` can be moved to StreamHandler/DataStream?